docs/workflow/debugging/coreclr/debugging-crossgen2.md

```diff
@@ -0,0 +1,65 @@
+  - Repro args will look like the following ``--singlemethodtypename "Internal.Runtime.CompilerServices.Unsafe" --singlemethodname As --singlemethodindex 2 --singlemethodgenericarg "System.Runtime.Intrinsics.Vector256`1[[System.SByte]]" --singlemethodgenericarg "System.Runtime.Intrinsics.Vector256`1[[System.Double]]"``
+
+- Since Crossgen2 is by default multi-threaded, it produces results fairly quickly even when compiling using a Debug variant of the JIT. In general, when debugging JIT issues we recommend using the debug JIT regardless of which environment caused a problem.
+
+- Crossgen2 supports nearly arbitrary cross-targetting, including OS and architecture cross targeting. The only restriction is that 32bit architecture cannot compile targetting a 64bit architecture. This allows the use of the debugging environment most convenient to the developer. In particular, if there is an issue which crosses the managed/native boundary, it is often convenient to debug using the mixed mode debugger on Windows X64.
+  - If the correct set of assemblies/command line arguments are passed to the compiler Crossgen2 should produce binary identical output on all platforms.
+  - The compiler does not check the OS/Architecture specified for input assemblies, which allows compiling using a non-architecture/OS matched version of the framework to target an arbitrary target. While this isn't useful for producing the diagnosing all issues, it can be cheaply used to identify the general behavior of a change on the full swath of supported architectures.
+
+Control compilation behavior by using the `--targetos` and `--targetarch` switches. The default behavior is to target the crossgen2's own OS/Arch pair, but all 64bit versions of crossgen2 are capable of targetting arbitrary OS/Arch combinations.
+At the time of writing the current supported sets of valid arguments are:
+| Command line arguments
+| --- |
+| `--targetos windows --targetarch x86` |
+| `--targetos windows --targetarch x64` |
+| `--targetos windows --targetarch arm` |
+|`--targetos windows --targetarch arm64` |
+|`--targetos linux --targetarch x64` |
+|`--targetos linux --targetarch arm` |
+|`--targetos linux --targetarch arm64` |
+|`--targetos osx --targetarch x64` |
+|`--targetos osx --targetarch arm64` |
+
+- Passing special jit behavior flags to the compiler is done via the `--codegenopt` switch. As an example to turn on tailcall loop optimizations and dump all code compiled use a pair of them like `--codegenopt NgenDump=* --codegenopt TailCallLoopOpt=1`.
+
+- When using the NgenDump feature of the JIT, disable parallelism as described above or specify a single method to be compiled. Otherwise, output from multiple functions will be interleaved and inscrutable.
```

Comment:
The JIT has a `COMPlus_JitStdOutFile` argument. Maybe the JIT should allow something like "{threadid}" in the name and crossgen2 tell us some unique compilation ID that we replace that with, to generate a unique filename.

Reply:
Crossgen2 has no reliable way to produce a unique compilation id for a given method. Unlike crossgen.exe method load order, and other details are sufficiently non-deterministic as to make that impossible without disabling parallelization. However, details like threadid are visible to the jit as it is just a normal bit of native code. The difficulty in reading the output comes from the mess of printf messages printed by the JIT on all the various threads. If you modify the JIT to have some sort of output that makes sense when the jit is run in parallel, please update this document with the technique.

Reply:
> Crossgen2 has no reliable way to produce a unique compilation id for a given method.

Interesting...

In this imagined scenario, it wouldn't need to be an id unique to the function (such that the same function always gets the same id).

Presumably the JIT could use some static global counter to incorporate into the name, as long as the JIT process itself doesn't get unloaded/re-loaded during compilation.